docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,292 @@
+can't create privileged pods AND regular users can't exec into that pod.
+
+Use cases 3:
+As a cluster administrator, I can allow a given namespace (or service account) to create privileged
+pods or to run root pods
+
+Use case 4:
+As a cluster administrator, I can allow a project administrator to control the security contexts of
+pods and service accounts within a project
+
+
+### Requirements
+
+1.  Provide a set of restrictions that a security context operates under a new object will be
+introduced called SecurityContextConstraints.
+1.  SecurityContextConstraints may also exist outside of a namespace.
+1.  A cluster default service account will be created for the default namespace and be available to
+admission controllers via configuration.
+1.  User information must be available to admission controllers.
+1.  Some authorizers may restrict a user’s ability to reference a service account.  Systems requiring
+the ability to secure service accounts on a user level will be able to add a policy that enables
+referencing specific service accounts themselves.
+1.  Admission control must validate the creation of Pods, PodTemplates, and Replication controllers
+against the allowed set of constraints.
+
+### Design
+
+#### Model
+
+The model for a security context constraint is a combination of the security context constraint
+itself and cluster wide strategies enforced by an allocator.  The separation of concerns falls into
+two distinct categories: permissions that the pod can request and runtime parameters the cluster
+admin controls.
+
+An example of a permission that can be requested is to run a pod as privileged.  This is set on a
+pod by pod basis.  A cluster wide setting may be the allocation of user ids that pods must run under.
+The cluster administrator may decide that no pod can run as the root user in their cluster and
+provide a block of UIDs that the allocator can use to distribute amongst namespaces.
+
+A security context constraint object exists in the root scope, outside of a namespace.  The
+security context constraints can reference users, groups, or a service account that is allowed
+to operate under the constraints.
+
+A one to one relationship will be maintained for a security context constraint and a service account.
+Users may be allowed to have many security context constraints and admission will create a unioned
+set of their highest allowed permissions when creating a pod (discussed in admission, below).
+
+To minimize the amount of objects that must be managed by a cluster administrator the following
+security context constraints objects can be provided as a default set:
+
+1.  User/Group Constraints - this is the default security context constraints for the cluster and is usually the
+most restrictive set of constraints.
+1.  Service Account Constraints - security context constraints for service accounts will be created that allows
```

Comment:
I don't think "The user should be required to have permissions to reference the service account in order to run a pod with the service account's security context constraints." belongs here... that would be the concern of the service account admission plugin. This system should only care what service account a pod references.